shims/src/peschke/shims/resource/ExceptionMapper.scala

```diff
@@ -0,0 +1,57 @@
+package peschke.shims.resource
+
+import cats.syntax.all._
+import cats.{ApplicativeError, Id}
+
+import scala.concurrent.Future
+import scala.util.Try
+import scala.util.control.NonFatal
+
+/** Something like [[cats.ApplicativeError]], specialized for our particular use
+  * case.
+  *
+  * Specifically: the need to easily adapt a [[Throwable]] into an `E`
+  */
+trait ExceptionMapper[F[_], E] {
+  def catchNonFatal[A](a:   => A):    F[A]
+  def catchNonFatalF[A](fa: => F[A]): F[A]
+}
+object ExceptionMapper {
+  def apply[F[_], E](implicit EM: ExceptionMapper[F, E]): EM.type = EM
+
+  def usingApplicativeError[F[_], E](mapErrorIfPossible: Throwable => Option[E])
+                                    (implicit AE:        ApplicativeError[F, E])
+    : ExceptionMapper[F, E] =
+    new ExceptionMapper[F, E] {
+      override def catchNonFatal[A](a: => A): F[A] = catchNonFatalF(a.pure[F])
+
+      override def catchNonFatalF[A](fa: => F[A]): F[A] =
+        try fa
+        catch {
+          case NonFatal(ex) =>
+            mapErrorIfPossible(ex).map(AE.raiseError[A]).getOrElse(throw ex)
+        }
+    }
+
+  implicit val doNothingForId: ExceptionMapper[Id, Throwable] =
+    new ExceptionMapper[Id, Throwable] {
+      override def catchNonFatal[A](a: => A): Id[A] = a
+
+      override def catchNonFatalF[A](fa: => Id[A]): Id[A] = fa
+    }
+
+  implicit val doNothingForTry: ExceptionMapper[Try, Throwable] =
```

Comment:
Why are you calling `doNothing`?

Reply:
Because it doesn't need to do anything, the exception stays an exception.